### PR TITLE
Non square Matrices

### PR DESCRIPTION
**Pull request by AlexMouton**
Hi Jutaro
I think this may be taking the code in a different direction than you intend, but I wanted to share.
I've made a number of [breaking] changes to allow for NonSquare matrices. 
Havn't delved into the STMatrix stuff at all so I imagine that will fail for NonSquare at present.
Any thoughts?

Thanks!
Alex

### CODE REVIEW

File: src/Data/Matrix.purs [hunk 1]
```diff
@@ -82,48 +124,77 @@ instance applyMat :: Apply (Mat s) where
...
+getElem m@(Mat l) i j = unsafePartial $ fromJust (l !! index m i j)
 
 -- | Scale a matrix by a given factor.
 --   Example:
 --
 -- >               ( 1 2 3 )   (  2  4  6 )
 -- >               ( 4 5 6 )   (  8 10 12 )
 -- > scaleMatrix 2 ( 7 8 9 ) = ( 14 16 18 )
-scaleMatrix ::  forall a s. (EuclideanRing a) => a -> Mat s a -> Mat s a
+scaleMatrix ::  forall r c a. (EuclideanRing a) => a -> Mat r c a -> Mat r c a
```
Comment (MonoidMusician) [inline src/Data/Matrix.purs:148]: could be relaxed to just `Semiring a`

File: src/Data/Matrix.purs [hunk 2]
```diff
@@ -82,48 +124,77 @@ instance applyMat :: Apply (Mat s) where
 -- >     ( 0 0 ... 1 0 )
 -- >   n ( 0 0 ... 0 1 )
 --
-identity' :: forall s.  (Sized s) => Mat s Number
+identity' :: forall r c. Sized r => Sized c => Mat r c Number
 identity' = generate \ i j -> if i == j then 1.0 else 0.0
```
Comment (MonoidMusician) [inline src/Data/Matrix.purs:120]: I would generalize this to any `Semiring` (which provides `one` and `zero`, including for `Int` and `Number`):
```
identity' :: forall a r c. Semiring a => Sized r => Sized c => Mat r c a
  identity' = generate \i j -> if i == j then one else zero
```

File: src/Data/Matrix.purs [hunk 3]
```diff
@@ -82,48 +124,77 @@ instance applyMat :: Apply (Mat s) where
...
+transpose m = generate $ \i j -> getElem m j i
+
+mulmatvec :: forall r c. Sized r => Sized c => Mat r c Number -> V.Vec c Number -> V.Vec r Number
+mulmatvec m v = V.Vec $ V.dot <$> rows m <*> [v]
+
+mulvecmat :: forall r c. Sized r => Sized c => V.Vec r Number -> Mat r c Number -> V.Vec c Number
+mulvecmat v m = V.Vec $ V.dot <$> [v] <*> columns m
+
+mulmatmat :: forall rl u cr. Sized rl => Sized u => Sized cr => Mat rl u Number -> Mat u cr Number -> Mat rl cr Number
+mulmatmat l r = fromColumns $ (mulmatvec l <$> columns r)
```
Comment (MonoidMusician) [inline src/Data/Matrix.purs:193]: These could also be generalized with `Semiring`, with upstream changes I guess ...

**Comment by AlexMouton** [comment]
Hello?

**Comment by jutaro** [comment]
Hi Alex,

actually I will drop purescript from my development tools.

So if you would be happy if you agree to take over a package.
But I will ask Michael first, because he co-developed some stuff.

My remaining interest is that backward compatibility is only broken for
very good reasons, and not
because some idea might seem fancier.

Sorry for the late answer.
Cheers
Jürgen

Alex Mouton <notifications@github.com> schrieb am Do., 4. Jan. 2018, 06:47:

**Comment by AlexMouton** [comment]
Hi Jurgen
Sounds reasonable. Let me know what Michael says.
I have 0.12 changes in process for -vector.  -matrix coming soon.

Thanks
Alex

**Comment by AlexMouton** [comment]
Hi Jurgen 
Just wondering if you have talked to michael about this repo?

Thanks
Alex

**Comment by jutaro** [comment]
Hi, Michaell will take over ownership next days.

**Pull request closed without merging**